File: 4.8-bc/go/stage-2-feature-first/app/notifications/handler.go
package notifications

import (
	"net/http"

	"github.com/go-chi/chi/v5"
	"github.com/google/uuid"

	"github.com/genkovich/claude-course-demos/4.8-bc/go/stage-2-feature-first/shared/httputil"
)

type Handler struct {
	svc *Service
}

func NewHandler(svc *Service) *Handler {
	return &Handler{svc: svc}
}

func (h *Handler) Register(r chi.Router) {
	r.Post("/notifications/test", h.sendTest)
}

type sendTestReq struct {
	UserID  string `json:"user_id"`
	Channel string `json:"channel"`
	Payload string `json:"payload"`
}

func (h *Handler) sendTest(w http.ResponseWriter, r *http.Request) {
	var req sendTestReq
	if err := httputil.DecodeBody(r, &req); err != nil {
		httputil.WriteJSON(w, http.StatusBadRequest, map[string]string{"error": "invalid_request"})
		return
	}
	uid, err := uuid.Parse(req.UserID)
	if err != nil {
		httputil.WriteJSON(w, http.StatusBadRequest, map[string]string{"error": "invalid_user_id"})
		return
	}
	if req.Channel == "" {
		req.Channel = "email"
	}
	if req.Payload == "" {
		req.Payload = "test notification"
	}
	n, err := h.svc.Send(r.Context(), uid, req.Channel, req.Payload)
	if err != nil {
		httputil.WriteError(w, err)
		return
	}
	httputil.WriteJSON(w, http.StatusOK, map[string]any{
		"notification_id": n.ID.String(),
		"channel":         n.Channel,
		"sent_at":         n.SentAt,
	})
}

File: 4.8-bc/py/stage-3-hexagonal/app/notifications/infra/http/handler.py
"""Notifications HTTP handler — POST /notifications/test."""
from __future__ import annotations

from datetime import datetime
from uuid import UUID

from fastapi import APIRouter, HTTPException, status
from pydantic import BaseModel

from app.notifications.app.service import Service
from app.notifications.domain.notification import Channel


class SendTestReq(BaseModel):
    user_id: str
    channel: str | None = None
    payload: str | None = None


class SendTestResp(BaseModel):
    notification_id: str
    channel: str
    sent_at: datetime | None


def build_router(svc: Service) -> APIRouter:
    router = APIRouter(tags=["notifications"])

    @router.post("/notifications/test", response_model=SendTestResp)
    async def send_test(req: SendTestReq) -> SendTestResp:
        try:
            uid = UUID(req.user_id)
        except (ValueError, TypeError) as exc:
            raise HTTPException(
                status_code=status.HTTP_400_BAD_REQUEST,
                detail="invalid_user_id",
            ) from exc
        channel = Channel(req.channel) if req.channel else Channel.EMAIL
        payload = req.payload or "test notification"
        n = await svc.send(uid, channel, payload)
        return SendTestResp(
            notification_id=str(n.id),
            channel=str(n.channel),
            sent_at=n.sent_at,
        )

    return router
